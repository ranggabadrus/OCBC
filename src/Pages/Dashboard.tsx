import {
  View,
  Text,
  TouchableOpacity,
  ActivityIndicator,
  FlatList,
  ToastAndroid,
} from 'react-native';
import React, {useEffect, useState} from 'react';
import {useNavigation} from '@react-navigation/native';
import {ScreenNavigationProp} from '../../App';
import Card from '../Components/Card';
import FloatButton from '../Components/FloatButton';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {useFormat} from '../Utils/useFormat';
import Modal from 'react-native-modal';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

const styles = require('../styles');

interface ModalProp {
  amount: number;
  description: string;
  person: {accountNo: string; accountHolder: string};
  transactionDate: any;
  transactionId: string;
  transactionType: string;
}

interface BalanceProp {
  accountNo: string;
  balance: number;
  status: string;
}

export default function Dashboard() {
  const [username, setUsername] = useState('');
  const [loading, setLoading] = useState(true);
  const [transaction, setTransaction] = useState([]);
  const [balance, setBalance] = useState<BalanceProp | null>(null);
  const navigation = useNavigation<ScreenNavigationProp>();
  const [isModal, setIsModal] = useState(false);
  const [modal, setModal] = useState<ModalProp | null>(null);

  const fetchData = async () => {
    const token = await AsyncStorage.getItem('token');
    try {
      const getData = await Promise.all([
        fetch('https://green-thumb-64168.uc.r.appspot.com/balance', {
          headers: {
            Authorization: `${token}`,
          },
        }),
        fetch('https://green-thumb-64168.uc.r.appspot.com/transactions', {
          headers: {
            Authorization: `${token}`,
          },
        }),
      ]);

      console.log('getData ', getData);
      if (getData[0].ok && getData[1].ok) {
        const res = await Promise.all(getData.map(async e => await e.json()));

        setBalance(res[0]);
        setTransaction(res[1].data);
      } else {
        ToastAndroid.show('Session expired', 7000);
        await AsyncStorage.removeItem('token');
        navigation.replace('Auth');
      }
      setLoading(false);
    } catch (error) {
      console.log('catch ', error);
    }
  };

  useEffect(() => {
    AsyncStorage.getItem('username').then(e => {
      if (e) {
        setUsername(e);
      }
    });

    fetchData();

    const unsubscribe = navigation.addListener('focus', () => {
      fetchData();
    });

    return () => {
      unsubscribe;
      setBalance(null);
      setTransaction([]);
    };
  }, [navigation]);

  if (loading) {
    return <ActivityIndicator color={'navy'} size="large" />;
  }

  console.log('trans ', modal);

  return (
    <View style={styles.flex1}>
      <TouchableOpacity
        onPress={async () => {
          await AsyncStorage.removeItem('token');
          navigation.replace('Auth');
        }}
        style={styles.logout}>
        <Text>Logout</Text>
      </TouchableOpacity>
      <View style={[styles.header]}>
        <Text style={styles.textMedium}>You have</Text>
        <Text style={[styles.textLarge, styles.textNavy]}>
          SGD${useFormat(balance?.balance)}
        </Text>
        <Text>Account No</Text>
        <Text style={styles.textMedium}>{balance?.accountNo}</Text>
        <Text>Account Holder</Text>
        <Text style={styles.textMedium}>{username}</Text>
      </View>
      <View style={styles.body}>
        <Text style={[styles.textMedium, styles.marginBottom10]}>
          Your transaction history
        </Text>
        <FlatList
          data={transaction}
          renderItem={({item}: any) => {
            const person =
              item.transactionType == 'transfer'
                ? item.receipient
                : item.sender;
            return (
              <TouchableOpacity
                onPress={() => {
                  setModal({...item, person});
                }}>
                <Card {...item} person={person} />
              </TouchableOpacity>
            );
          }}
          showsVerticalScrollIndicator={false}
          keyExtractor={(_, index) => `${index}`}
          contentContainerStyle={{paddingBottom: 50}}
        />
      </View>
      <FloatButton
        task={() => navigation.navigate('Transfer')}
        text={'Make Transfer'}
      />

      <Modal
        isVisible={modal ? true : false}
        onBackdropPress={() => setModal(null)}
        onBackButtonPress={() => setModal(null)}>
        <View style={styles.modalContainer}>
          <Text style={[styles.textMedium, styles.textCenter]}>
            Details Transaction
          </Text>
          <Text></Text>
          <Text>
            Transaction type:{' '}
            <Text style={[styles.bold, styles.textBlack]}>
              {modal?.transactionType.toUpperCase()}
            </Text>
          </Text>
          <Text>
            {new Date(modal?.transactionDate).toLocaleString()}{' '}
            <Icon name="clock" size={20} />
          </Text>

          <Text></Text>
          <Text style={[styles.bold, styles.textBlack]}>SENDER</Text>
          <View style={styles.row}>
            <Text style={styles.text}>Name:</Text>
            <Text style={styles.bold}>: {username}</Text>
          </View>
          <View style={styles.row}>
            <Text style={styles.text}>Account:</Text>
            <Text style={styles.bold}>: {balance?.accountNo}</Text>
          </View>
          <View style={styles.row}>
            <Text style={styles.text}>Amount: </Text>
            <Text style={styles.bold}>: SGD${useFormat(modal?.amount)}</Text>
          </View>
          <View style={styles.row}>
            <Text style={styles.text}>Description: </Text>
            <Text style={styles.bold}>: {modal?.description}</Text>
          </View>

          <Text></Text>
          <Text style={[styles.bold, styles.textBlack]}>RECEIPIENT</Text>
          <View style={styles.row}>
            <Text style={styles.text}>Name</Text>
            <Text style={styles.bold}>: {modal?.person.accountHolder}</Text>
          </View>
          <View style={styles.row}>
            <Text style={styles.text}>Account</Text>
            <Text style={styles.bold}>: {modal?.person.accountNo}</Text>
          </View>

          <Text>{}</Text>
          <View style={styles.row}>
            <Text style={styles.text}>Transaction Id:</Text>
            <Text style={styles.bold}>: {modal?.transactionId}</Text>
          </View>
          <TouchableOpacity
            style={styles.close}
            onPress={() => {
              setModal(null);
            }}>
            <Text style={styles.textWhite}>Close</Text>
          </TouchableOpacity>
        </View>
      </Modal>
    </View>
  );
}
