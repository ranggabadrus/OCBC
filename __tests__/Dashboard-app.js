import React, {useRef} from 'react';
import {
  render,
  fireEvent,
  waitFor,
  act,
  cleanup,
} from '@testing-library/react-native';
import {useNavigation} from '@react-navigation/native';
import Dashboard from '../src/Pages/Dashboard';
import AsyncStorage from '@react-native-async-storage/async-storage';

jest.mock('react', () => {
  const actualNav = jest.requireActual('react');
  return {
    ...actualNav,
    useRef: jest.fn(),
  };
});

describe('Dashboard Page ', () => {
  beforeEach(() => {
    fetch.mockResponses(
      [
        JSON.stringify({
          accountNo: '12345',
          balance: 111111,
          status: 'success',
        }),
      ],
      [
        JSON.stringify({
          status: 'success',
          data: [
            {
              transactionId: '622ba782f3aa395d71354900',
              amount: 100000,
              transactionDate: '2022-03-11T19:48:18.249Z',
              description: 'Default money-in transaction',
              transactionType: 'received',
              sender: {
                accountNo: '1111-111-1111',
                accountHolder: 'Jane',
              },
            },
          ],
        }),
      ],
    );
  });

  it('render dashboard with fetched data', async () => {
    const navigation = jest.fn();
    useNavigation.mockImplementation(() => ({
      navigate: navigation,
      addListener: navigation,
      replace: navigation,
    }));

    const ref = jest.fn();

    useRef.mockImplementation(() => ({
      current: ref,
    }));

    const {toJSON, getByTestId, findByTestId, findAllByTestId} = render(
      <Dashboard />,
    );

    await waitFor(async () => {
      const balance = getByTestId('balance', {}, {timeout: 2000});
      expect(balance.props.children).not.toEqual('NaN');
      const accountNo = getByTestId('accountNo', {}, {timeout: 2000});
      expect(accountNo.props.children).toEqual('12345');
      const transaction = getByTestId('transaction', {}, {timeout: 2000});
      expect(transaction.props.data).toHaveLength(1);
    });

    // expect(toJSON()).toMatchSnapshot();
  });

  it('from dashboard go to transfer page', async () => {
    const navigation = jest.fn();
    useNavigation.mockImplementation(() => ({
      navigate: navigation,
      addListener: navigation,
      replace: navigation,
    }));

    const ref = jest.fn();

    useRef.mockImplementation(() => ({
      current: ref,
    }));

    const {toJSON, getByTestId, findByTestId, findAllByTestId} = render(
      <Dashboard />,
    );

    await waitFor(async () => {
      const transfer = getByTestId('goToTransfer');
      fireEvent.press(transfer);
      expect(navigation).toHaveBeenCalledWith('Transfer');
    });

    expect(toJSON()).toMatchSnapshot();
  });

  it('open modal for detail transaction', async () => {
    const navigation = jest.fn();
    useNavigation.mockImplementation(() => ({
      navigate: navigation,
      addListener: navigation,
      replace: navigation,
    }));

    const {toJSON, getByTestId, findByTestId, findAllByTestId} = render(
      <Dashboard />,
    );

    await waitFor(async () => {
      const button = getByTestId('detail');
      fireEvent.press(button);
      const detailModal = getByTestId('detailModal');
      expect(detailModal.props.visible).toEqual(true);
    });

    // expect(toJSON()).toMatchSnapshot();
  });
});
