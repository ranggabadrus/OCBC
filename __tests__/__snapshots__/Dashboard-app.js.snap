// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Page  from dashboard go to transfer page 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "position": "absolute",
        "top": 0,
      }
    }
  >
    <RNSVGSvgView
      bbHeight={500}
      bbWidth={750}
      focusable={false}
      height={500}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 500,
            "width": 750,
          },
        ]
      }
      width={750}
    >
      <RNSVGGroup>
        <RNSVGCircle
          cx={-50}
          cy={-300}
          fill={4278190208}
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          r={500}
          stroke={4278190208}
          strokeWidth="2"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": "flex-end",
        "backgroundColor": "#f2f2f2",
        "margin": 10,
        "transform": Array [
          Object {
            "translateX": [MockFunction],
          },
        ],
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "gray",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        ó°—½
      </Text>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginLeft": 5,
            "opacity": 1,
          }
        }
      >
        <Text>
          Logout
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "backgroundColor": "#fff",
          "borderBottomEndRadius": 50,
          "borderColor": "navy",
          "borderTopEndRadius": 50,
          "borderWidth": 2,
          "elevation": 20,
          "opacity": 0.9,
          "padding": 20,
          "width": 600,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
      testID="money"
    >
      You have
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 30,
            "fontWeight": "bold",
          },
        ]
      }
    >
      SGD$
      <Text
        testID="balance"
      >
        111,111.00
      </Text>
    </Text>
    <Text>
      Account No
    </Text>
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
      testID="accountNo"
    >
      12345
    </Text>
    <Text>
      Account Holder
    </Text>
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "margin": 20,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 20,
            "fontWeight": "bold",
          },
          Object {
            "marginBottom": 10,
          },
        ]
      }
    >
      Your transaction history
    </Text>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "paddingBottom": 50,
        }
      }
      data={
        Array [
          Object {
            "amount": 100000,
            "description": "Default money-in transaction",
            "sender": Object {
              "accountHolder": "Jane",
              "accountNo": "1111-111-1111",
            },
            "transactionDate": "2022-03-11T19:48:18.249Z",
            "transactionId": "622ba782f3aa395d71354900",
            "transactionType": "received",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      testID="transaction"
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="detail"
          >
            <View
              style={
                Object {
                  "backgroundColor": "#fff",
                  "borderColor": "#eaeaea",
                  "borderRadius": 20,
                  "borderWidth": 2,
                  "height": 100,
                  "marginVertical": 5,
                  "padding": 10,
                }
              }
            >
              <Text>
                3/12/2022, 2:48:18 AM
              </Text>
              <View
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                    },
                    Object {
                      "justifyContent": "space-between",
                    },
                  ]
                }
              >
                <View>
                  <Text>
                    Jane
                  </Text>
                  <Text>
                    1111-111-1111
                  </Text>
                </View>
                <View>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#20cb9d",
                        },
                        Object {
                          "fontWeight": "bold",
                        },
                      ]
                    }
                  >
                    +SGD$
                    100,000.00
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "left": 0,
        "marginBottom": 20,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "black",
          "borderRadius": 50,
          "opacity": 1,
          "paddingHorizontal": 50,
          "paddingVertical": 10,
        }
      }
      testID="goToTransfer"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
              "fontSize": 20,
              "fontWeight": "bold",
            },
            Object {
              "color": "#fff",
            },
          ]
        }
      >
        Make Transfer
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "bottom": 0,
        "position": "absolute",
        "zIndex": -1,
      }
    }
  >
    <RNSVGSvgView
      bbHeight={500}
      bbWidth={750}
      focusable={false}
      height={500}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 500,
            "width": 750,
          },
        ]
      }
      width={750}
    >
      <RNSVGGroup>
        <RNSVGCircle
          cx={450}
          cy={800}
          fill={4278190208}
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          r={500}
          stroke={4278190208}
          strokeWidth="2"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    testID="detailModal"
    transparent={true}
    visible={false}
  />
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  />
</View>
`;
